ENH: Change from Debug to Warning macro.

zivy opened the pull request:
When non DICOM and non standard ITK tags are in the metadata dictionary, issue a warning. This addresses the common mistake of using the comma as a separator in the group-element key instead of the pipe, which is what ITK/GDCM uses. These metadata dictionary entries were silently ignored. The warning should provide the user with a hint about what is happening.

dzenanz commented on Modules/IO/GDCM/src/itkGDCMImageIO.cxx:
```diff
@@ -1008,7 +1008,9 @@ GDCMImageIO::Write(const void * buffer)
       else
       {
-        itkDebugMacro("GDCMImageIO: non-DICOM and non-ITK standard key = " << key);
+        itkWarningMacro(
+          "ignoring non-DICOM and non-ITK standard key (DICOM key structure is group|element with pipe separator) = "
+          << key);
```
Should there be a way to suppress this warning? If someone knows that not their entire metadata dictionary will be written, but don't want a warning (and don't want to clean it up)?

N-Dekker replied:
By the way, I wonder why the common `(XXXX,XXXX)` format isn't supported anyway 🤷

blowekamp replied:
GDCM has three methods to parse the tags, only one is currently being used. 
https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/ThirdParty/GDCM/src/gdcm/Source/DataStructureAndEncodingDefinition/gdcmTag.h#L244-L259

blowekamp replied:
I'd also recommend 1 warning per file not per tag. That is something like "the following tags are non-standard and are being ignored: ...."

N-Dekker replied:
Thanks @blowekamp! That would suggest changing the following line of code: 
https://github.com/InsightSoftwareConsortium/ITK/blob/e93fa19d909ef06982d8d25b347d575a1911aee9/Modules/IO/GDCM/src/itkGDCMImageIO.cxx#L875

To something like:

```cpp

   bool b = tag.ReadFromPipeSeparatedString(key.c_str()) ||
       tag.ReadFromCommaSeparatedString(key.c_str()); 

```

Right?

blowekamp replied:
Or maybe a state in the reader to switch between them?

zivy replied:
Per @N-Dekker suggestion, my preference is that both comma and pipe separated strings be accepted without a switch controlling which to use. 

This addresses the issue encountered by most users who copy paste the tags from some other source. The comma separator is the commonly encountered one outside of the ITK ecosystem. Also, after making this change the warning can be reverted back to debug. 

If there's agreement on this, I'll change the PR to match.

blowekamp replied:
There would still be the issue of the GDCM not supporting parenthesis, which are commonly used too.

dzenanz replied:
I would still like to keep the warning, but switch it to one per file.

N-Dekker replied:
Thanks @zivy Of course, if we support both pipe and comma separators, there _might be_ a tiny little chance that the very same DICOM tag is represented in two ways `XXXX,XXXX` and `XXXX|XXXX`, in one and the same file. But I guess that would be a very rare mistake, that we may ignore. Right?

zivy replied:
I will change the PR so that the warning is per file and not per tag and that both pipe and comma are acceptable separators. 
@N-Dekker, you raised a valid corner case which I'd like to ignore for now given the expected rarity (will add it to the documentation instead of directly dealing with it and revisit in the future).

blowekamp replied:
Please add testing of the new functionality.

zivy commented:
Updated and force pushed. Writing supports both comma and pipe separators and added testing. Non-DICOM and non-ITK metadata in dictionary trigger a single warning per file which lists all these unexpected dictionary keys.
